lib/view_component/collection.rb

```diff
@@ -29,10 +29,22 @@ def collection_variable(object)
       end
     end
 
+    def collection_parameter_name
+      @component.collection_parameter_name
+    end
+
+    def collection_counter_parameter_name
+      "#{collection_parameter_name}_counter".to_sym
+    end
+
     def component_options(item, iterator)
-      as = @component.collection_parameter_name
+      @options.merge!(collection_parameter_name => item)
+      @options.merge!(collection_counter_parameter_name => iterator.index) if counter_argument_present?
+      @options
+    end
 
-      @options.merge(as => item, "#{as}_counter": iterator.index)
+    def counter_argument_present?
+      !!@component.instance_method(:initialize).parameters.dup.map(&:second).include?(collection_counter_parameter_name)
```

Comment:
Love this approach.

Nits to pick:

* Do we need the "!!" `include?` always returns a boolean not nil right?
* I don't think we need dup. given we map
* should we memoize all these properties?

Reply:
Yeah, I think it won't hurt to memoize and you are correct about `include?` always returning a boolean.